### PR TITLE
Enum bugs

### PR DESCRIPTION
**Pull request by anatoliykmetyuk**
I have included 3 issues in this pull request. Below there are links to the corresponding commits.

1. [Bug fix](https://github.com/anatoliykmetyuk/scalaxb/compare/40a42499f414f62c38b008c368a946e02aa5372f...anatoliykmetyuk:e087df8aeedb0ee167b4af2d196aa620413e1db0) for name clashes between enum value bindings and Scala predef values
2. [Bug fix](https://github.com/anatoliykmetyuk/scalaxb/commit/0f4e435b5762eef1022437e3e398661512664466) for data types that are not taken into account for Enums
3. [Bug fix](https://github.com/eed3si9n/scalaxb/commit/46078d40c306cef882a3be027d04c09c7504a4e4) for the protocol package to always have a value

Also, some tests were [updated](https://github.com/eed3si9n/scalaxb/compare/46078d40c306cef882a3be027d04c09c7504a4e4...002e1002260144df3eaa809004b51495e70c830f) accordingly.

I wanted to grain the work even finer, but these issues are tightly dependent one on another. For the (2) bug, we use `fromXml` from the generated binding classes, which requires the protocol imports. So we need a guarantee that the protocol will have a determinable package name defined, which is introduced in (3).

Also, `enumName` function added to solve (1) is used in (2).

Below I'll copy-paste the info on the bugs from my previous pull request:

## Name clashes between enum value bindings and Scala predef values
Consider this snippet from OpenOffice's `sml.xsd`:

```xml
  <xsd:simpleType name="ST_DdeValueType">
    <xsd:restriction base="xsd:string">
      <xsd:enumeration value="nil"/>
      <xsd:enumeration value="b"/>
      <xsd:enumeration value="n"/>
      <xsd:enumeration value="e"/>
      <xsd:enumeration value="str"/>
    </xsd:restriction>
  </xsd:simpleType>
```

On line 3, the enum binding object for that value will be named Nil, which clashes with Scala's predefined Nil. We fixed that by prefixing the object name with "EnumValue_" if this name clashes with a Scala keyword or a commonly used name.

## Data types are taken into account for Enums
Consider the following snippet from `dml-main.xsd`:

```xml
  <xsd:simpleType name="ST_PitchFamily">
   <xsd:restriction base="xsd:byte">
     <xsd:enumeration value="00"/>
     <xsd:enumeration value="01"/>
     <xsd:enumeration value="02"/>
     <xsd:enumeration value="16"/>
     <xsd:enumeration value="17"/>
     <xsd:enumeration value="18"/>
     <xsd:enumeration value="32"/>
     <xsd:enumeration value="33"/>
     <xsd:enumeration value="34"/>
     <xsd:enumeration value="48"/>
     <xsd:enumeration value="49"/>
     <xsd:enumeration value="50"/>
     <xsd:enumeration value="64"/>
     <xsd:enumeration value="65"/>
     <xsd:enumeration value="66"/>
     <xsd:enumeration value="80"/>
     <xsd:enumeration value="81"/>
     <xsd:enumeration value="82"/>
   </xsd:restriction>
   </xsd:simpleType>
   
   <!-- ... -->
<xsd:attribute name="pitchFamily" type="ST_PitchFamily" use="optional" default="0"/>
```

Lines of interest are 3 and 25. The type of `ST_PitchFamily` is `byte`, so it is reasonable to expect `00` and `0` to be the same value. But in ScalaXB, `0` will throw an exception when you try to convert it to `ST_PitchFamily`, since ScalaXB treats all its enum values as strings.

After our changes, the code generated for `ST_PitchFamily` is this:

```scala
object ST_PitchFamily {
  def fromString(value: String, scope: scala.xml.NamespaceBinding): ST_PitchFamily = scalaxb.fromXML[Byte](scala.xml.Text(value)) match {
    case x: Byte if x == scalaxb.fromXML[Byte](scala.xml.Text("00")) => Number00
    case x: Byte if x == scalaxb.fromXML[Byte](scala.xml.Text("01")) => Number01
	// ... and so on
  }
}
```

That is, we convert the passed `value` string to the type declared in the enum, `Byte` in this case. And all the predefined enum values are also converted to this type. And the results of the conversion are compared. `00` and `0` are both converted to a `Byte` `0`, so everything works.

Previously, if I recall correctly, the code above was something like this:

```scala
object ST_PitchFamily {
  def fromString(value: String, scope: scala.xml.NamespaceBinding): ST_PitchFamily = value match {
    case "00" => Number00
    case "01" => Number01
	// ... and so on
  }
}
```

That obviously didn't work with "0".

## Protocol package is always guaranteed to be set
See issue #382. Protocol package can't be empty, since protocol is generated as "object `package`". It is reasonable that it should always be defined.

Previously, if no packages were specified by the user, the protocol was generated without a package. If some packages were specified by the user, but not for the protocol, the protocol occupied some package from the list specified by the user. This is a counterintuitive behaviour, since the protocol has the entities for all the packages, so its own package must be something more global.

We made the protocol to always have a meaningful package name. The package name selection goes as follows:

1. If `--protocol-package` is set, use it
2. Else, if `-p`, (`--default-package`) is set, use it (i.e. use the default package name)
3. Else, use a reasonable default - `scalaxb.package`

### CODE REVIEW

File: cli/src/main/scala/scalaxb/compiler/Config.scala [hunk 1]
```diff
@@ -85,15 +90,15 @@ object Config {
   val defaultOutdir = Outdir(new File("."))
   val defaultWrappedComplexTypes = WrappedComplexTypes(Nil)
   val defaultProtocolFileName = ProtocolFileName("xmlprotocol.scala")
-  val defaultProtocolPackageName = ProtocolPackageName(None)
+  val defaultProtocolPackageName = "scalaxb.protocol"
```
Comment (eed3si9n) [inline cli/src/main/scala/scalaxb/compiler/Config.scala]: I don't want `scalaxb` package used for the code that we didn't write. The default package should be `"generated"`.

File: sbt-scalaxb/src/main/scala/sbtscalaxb/Plugin.scala [hunk 2]
```diff
@@ -30,7 +30,7 @@ object Plugin extends sbt.Plugin {
     lazy val packageDir       = SettingKey[Boolean]("scalaxb-package-dir")
     lazy val generateRuntime  = SettingKey[Boolean]("scalaxb-generate-runtime")
     lazy val protocolFileName = SettingKey[String]("scalaxb-protocol-file-name")
-    lazy val protocolPackageName  = SettingKey[Option[String]]("scalaxb-protocol-package-name")
+    lazy val protocolPackageName  = SettingKey[String]("scalaxb-protocol-package-name")
```
Comment (eed3si9n) [inline sbt-scalaxb/src/main/scala/sbtscalaxb/Plugin.scala]: Making this change would break every sbt-scalaxb user's build, which I don't think is worth it since existing users would likely have a `Some("foo")` set already, and they don't gain anything from it.

**Comment by eed3si9n** [comment]
I'm ok with the enum bug fixes, but I don't think the protocol package thing is right if that changes the user experience of the emitted code just to catch some corner case.
Could you split 46078d4 onwards to another pull request?

File: integration/src/test/scala/XmlSchemaTest.scala [hunk 3]
```diff
@@ -15,6 +15,7 @@ object XmlSchemaTest extends TestBase {
   lazy val generated = module.process(inFile, config)
   "XMLSchema.scala file must compile so that Schema can be used" in {
     (List("import scalaxb._",
+      "import scalaxb.protocol._",
```
Comment (eed3si9n) [inline integration/src/test/scala/XmlSchemaTest.scala]: If you had to do change this here, it implies that all existing scalaxb users would have to do the same, and it's actually a nonsensical value (`scalaxb.protocol`). Shouldn't the protocol package fall back to the package used for case classes?

**Comment by anatoliykmetyuk** [comment]
> Shouldn't the protocol package fall back to the package used for case classes?

That's the way it [behaves](https://github.com/eed3si9n/scalaxb/blob/master/cli/src/main/scala/scalaxb/compiler/Module.scala#L329) now. This works fine in case of only one default package, but if, say, you have `package a` for one schema and `package b` for another schema, the protocol will end up in one of them. Say `a`. Then, if you only need to use package `b` from your code, it will rise eyebrows if you need to `import a._`.

Further, if you generate a separate package for each schema when you have a dozen of different schemas, it will take you some time to locate where the protocol was actually generated. I don't think it is possible to hide the protocol from the end user in all cases, unless you generate the protocol in fragments for every package defined.

**Comment by anatoliykmetyuk** [comment]
> I'm ok with the enum bug fixes, but I don't think the protocol package thing is right if that changes the user experience of the emitted code just to catch some corner case.
Could you split 46078d4 onwards to another pull request?

The enum bug fixes depend on the knowledge of the protocol package. Protocol is [imported](https://github.com/eed3si9n/scalaxb/commit/46078d40c306cef882a3be027d04c09c7504a4e4#diff-4376ad4608a978c1cca8ea57a58b9856R48) into each case-class file so that it is possible to call `fromXML` there. It should be possible to infer the protocol name from that line to enable the current fixes.

**Comment by eed3si9n** [comment]
> but if, say, you have package a for one schema and package b for another schema...

Every time we make breaking changes that force the user to change the usage code, it incurs time/money on their part. Therefore we should minimize anything that can disrupt the usage code.

Also note that someone can set up scalaxb to generate multiple configurations of schema sets with separate protocols, for example for different SOAP web services. My guess is that more than half of the users would just deal with one namespace/package at a time. We can let the user specify a particular package name for the protocol, but the default should behaves as it's behaving now so we break less.

**Comment by eed3si9n** [comment]
> The enum bug fixes depend on the knowledge of the protocol package. Protocol is imported into each case-class file so that it is possible to call fromXML there. It should be possible to infer the protocol name from that line to enable the current fixes.

Hmm.. I've always tried to make the separation of datatypes and the XML protocol, so the protocol depend on the datatypes but never other way around. We should actually introduce [Read typeclass](http://eed3si9n.com/scopt3).

**Comment by anatoliykmetyuk** [comment]
Is Read typeclasses a DRY approach? They seem to be `String => T` functions, while `XMLFormat`s are `scala.elem.Node => T`, it can duplicate a lot of functionality if we use both. They also seem to be of the same nature - the parsing logic, so if we want to keep the data model apart from one, we'll want to keep it apart from the other.

Why is the separation between the parsing logic and the data model needed? What harm will it cause if the data model is able to use the parsing logic?

**Comment by eed3si9n** [comment]
> Why is the separation between the parsing logic and the data model needed? What harm will it cause if the data model is able to use the parsing logic?

I'm trying to simplify the source dependencies. It makes it easier to conceptualize the separation of responsibilities, improves compilation speed, and avoids circular dependencies.

 > Is Read typeclasses a DRY approach? They seem to be String => T functions, while XMLFormats are scala.elem.Node => T, it can duplicate a lot of functionality if we use both. They also seem to be of the same nature - the parsing logic, so if we want to keep the data model apart from one, we'll want to keep it apart from the other.

That's a good point. My adding `fromString` to the companion object probably was the beginning of the mistake. With your approach the enum parsing is closer to the generation of normal `XMLFormat`, so perhaps we should try generating `XMLFormat` and derive the `fromString` function automatically.

Here's what it could look like:

```scala
trait XMinOccurs

object XMinOccurs {
  def fromString(value: String, scope: scala.xml.NamespaceBinding)(implicit ev: XMLFormat[XMinOccurs]): XMinOccurs =
    scalaxb.deriveFromString[XMinOccurs]
}

case object X0 extends XMinOccurs { override def toString = "0" }
case object X1 extends XMinOccurs { override def toString = "1" }
```

....


```scala
  trait DefaultXmlschemaXMinOccursFormat extends scalaxb.XMLFormat[xmlschema.XMinOccurs] {
    val targetNamespace: Option[String] = Some("http://www.w3.org/2001/XMLSchema")
    
    // put fromString here
    def fromString ... 

    def reads(seq: scala.xml.NodeSeq, stack: List[scalaxb.ElemName]): Either[String, xmlschema.XMinOccurs] = seq match {
      case elem: scala.xml.Elem => Right(fromString(elem.text, elem.scope))
      case _ => Right(fromString(seq.text, scala.xml.TopScope))
    }
    
    def writes(__obj: xmlschema.XMinOccurs, __namespace: Option[String], __elementLabel: Option[String],
        __scope: scala.xml.NamespaceBinding, __typeAttribute: Boolean): scala.xml.NodeSeq =
      scala.xml.Elem(scalaxb.Helper.getPrefix(__namespace, __scope).orNull, 
        __elementLabel getOrElse { sys.error("missing element label.") },
        scala.xml.Null, __scope, true, scala.xml.Text(__obj.toString))
  }
```

**Comment by anatoliykmetyuk** [comment]
I see, looks good. Though we'll need to use another trait instead of XMLFormat in the implicit position, since it doesn't have the `fromString` method, I believe. I'll try this approach.

**Comment by anatoliykmetyuk** [comment]
Though we'llbe able to run `reads` from `deriveFromString`, so XMLFormat should be OK.
